wdl-grammar/src/v1/lint/version_declaration_placement.rs

```diff
@@ -0,0 +1,214 @@
+        Code::try_new(code::Kind::Error, Version::V1, 9).unwrap()
+    }
+
+    fn group(&self) -> Group {
+        Group::Spacing
+    }
+
+    fn check(&self, tree: &Pair<'_, v1::Rule>) -> lint::Result {
+        let mut warnings = VecDeque::new();
+
+        // Optionally consume comment nodes, if found, allow exactly one empty line
+        // between the comment and the version declaration.
+        let mut comment = 0;
+        let mut newline = 0;
+        let mut anything_else = 0;
+
+        // This will never get used. Validation rules require a version statement.
+        let mut location: Location = Location::Unplaced;
+
+        for node in tree.clone().into_inner() {
+            match node.as_rule() {
+                v1::Rule::COMMENT => {
+                    comment += 1;
+                }
+                v1::Rule::WHITESPACE => {
+                    if node.as_str() == "\n" {
+                        newline += 1;
+                    } else {
+                        anything_else += 1;
+                    }
+                }
+                v1::Rule::version => {
+                    location = Location::try_from(node.as_span()).map_err(lint::Error::Location)?;
+                    break;
+                }
+                _ => {
+                    anything_else += 1;
+                }
+            }
+        }
+
+        // If anything other than version, comment, and newlines detected.
+        if (anything_else > 0)
+            // If comments detected, there should be one empty line between comments and version.
+            || (comment > 0 && newline != (comment + 1))
+            // If no comments detected, version should be the first line.
+            || (comment == 0 && newline != 0)
+        {
+            warnings.push_back(self.misplaced_version(location))
+        }
+
+        match warnings.pop_front() {
+            Some(front) => {
+                let mut results = NonEmpty::new(front);
+                results.extend(warnings);
+                Ok(Some(results))
+            }
+            None => Ok(None),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pest::Parser as _;
+    use wdl_core::concern::lint::Rule as _;
+
+    use super::*;
+    use crate::v1::parse::Parser;
+    use crate::v1::Rule;
+
+    #[test]
+    fn it_catches_missing_newline() -> Result<(), Box<dyn std::error::Error>> {
+        let tree = Parser::parse(
+            Rule::document,
+            r#"## Header comment
+version 1.0"#,
+        )?
+        .next()
+        .unwrap();
+
+        let warnings = VersionDeclarationPlacement.check(&tree)?.unwrap();
+
+        assert_eq!(warnings.len(), 1);
+        assert_eq!(
+            warnings.first().to_string(),
+            "[v1::E009::Spacing/Medium] Improperly placed version declaration (2:1-2:12)"
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn it_catches_leading_newline_with_comment() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
IMO, either this rule needs to be renamed to be broader, or this case should be caught by a different rule. I'm not sure which, but this test case feels wrong to me. The leading newline we're complaining about here seems too far away from the version line to be part of the same rule. Unless that rule is broader, something like `document_preamble`? I think referring to the header+version declaration as a "document preamble" could work.

Reply:
The test was wrong. It was meant to test the case where extraneous whitespace existed, but no whitespace between `COMMENT` and `VERSION` lines.

Reply:
Ok then I think I'm confused. I can think of two missing test cases, which I think if added would clear up the intended behavior.

It's not clear if this is intended to pass or not:
```

#comment

version 1.0
```
I think if it _is not_ intended to pass, the name/text of the rule should be broader and include the whole "preamble" as I'm calling it. If it _is_ intended to pass, then this PR is fine but we need another rule for "no leading whitespace".

And lastly, just a sanity check the handles the most basic case correctly:
```
version 1.0
```

Reply:
I do check for leading whitespace in `it_catches_leading_newline` without a `COMMENT`, since that would be a version declaration placement problem. Reviewing the list of lint rules, I don't see one that would check for whitespace preceding a `COMMENT` line, so maybe it would be best to expand this rule to handle the full document "preamble".
- No leading whitespace
- Optional comment(s) followed by exactly one empty line
- a version declaration

Reply:
I think handling all three of those bullet points with one rule makes the most sense. That would become a `preamble` rule as opposed to a `version_declaration` rule IMO.

Reply:
I've renamed to `document_preamble` and implemented those checks.